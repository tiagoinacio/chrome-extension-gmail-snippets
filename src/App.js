import React from 'react';

const App = () => (
    <iframe
        src='https://email-snippets.herokuapp.com'>
        Your browser does not support iframes.
    </iframe>
);

export default App;
