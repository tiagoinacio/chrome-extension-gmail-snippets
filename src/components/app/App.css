.iframe {
  height: 100%;
  width: 100%;
  border-radius: 4px;
  border: 1px solid #c3c3c3;
  box-shadow: 0 1px 2px 0 rgba(60,64,67,0.302), 0 2px 6px 2px rgba(60,64,67,0.149);
}

.container {
  height: 100%;
}

.loader {
  position: absolute;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
  display: flex;
  background-color: rgba(255, 255, 255, 0.7);
}

.loader:after {
  content: " ";
  display: block;
  width: 46px;
  height: 46px;
  margin: 1px;
  border-radius: 50%;
  border: 5px solid #c3c3c3;
  border-color: #c3c3c3 transparent #c3c3c3 transparent;
  animation: loader 1.2s linear infinite;
}
@keyframes loader {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
