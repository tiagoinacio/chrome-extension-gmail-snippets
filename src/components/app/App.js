import React, { Component } from 'react';
import PropTypes from 'prop-types';
import './App.css';

class App extends Component {
    static propTypes = {
        composeView: PropTypes.object.isRequired,
    };

    state = {
        isIframeReady: false,
    };

    componentDidMount() {
        window.addEventListener('message', this.handleMessageReceived, false);
    }

    render() {
        return (
            <div className="container">
                { !this.state.isIframeReady && <div className="loader"></div> }
                <iframe
                    ref={(iframe) => { this.iframe = iframe; }}
                    onLoad={this.onIframeReady}
                    className="iframe"
                    src='https://email-snippets.herokuapp.com'>
                    Your browser does not support iframes.
                </iframe>
            </div>
        );
    }

    componentWillUnmount() {
        window.removeEventListener('message', this.handleMessageReceived);
    }

    handleMessageReceived = (message) => {
        const { composeView } = this.props;

        if (message.origin === 'https://email-snippets.herokuapp.com') {
            composeView.insertTextIntoBodyAtCursor(message.data.template);
        }
    }

    onIframeReady = () => {
        this.setState({
            isIframeReady: true,
        });
    }
}

export default App;
