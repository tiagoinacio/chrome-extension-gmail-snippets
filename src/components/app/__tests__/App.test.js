import { mount } from 'enzyme';
import App from '..';
import React from 'react';

describe('App', () => {
    let composeView;
    let app;

    beforeEach(() => {
        global.window.removeEventListener = jest.fn();
        global.window.addEventListener = jest.fn();

        composeView = {
            insertTextIntoBodyAtCursor: jest.fn(),
        };

        app = mount(
            <App composeView={ composeView }/>,
        );
    });

    describe('while iframe is not loaded', () => {
        it('should render the loader', () => {
            expect(app).toMatchSnapshot();
        });
    });

    describe('when iframe is loaded', () => {
        it('should not render the loader', () => {
            // See this issue: https://github.com/airbnb/enzyme/issues/566
            app.findWhere(node => node.getDOMNode() === app.instance().iframe).simulate('load');

            expect(app).toMatchSnapshot();
        });
    });

    describe('#componentDidMount', () => {
        it('should add the event listener "message"', () => {
            expect(global.window.addEventListener).toHaveBeenCalledWith(
                'message',
                expect.any(Function),
                false,
            );
        });

        describe('on message received', () => {
            describe('when origin matches our webserver', () => {
                beforeEach(() => {
                    global.window.addEventListener = jest.fn((_, cb) => cb({
                        origin: 'https://email-snippets.herokuapp.com',
                        data: {
                            template: 'Test template',
                        },
                    }));

                    app = mount(
                        <App composeView={ composeView }/>,
                    );
                });

                it('should insert the template into the view', () => {
                    expect(composeView.insertTextIntoBodyAtCursor).toHaveBeenCalledWith('Test template');
                });
            });

            describe('when origin does not match our webserver', () => {
                beforeEach(() => {
                    global.window.addEventListener = jest.fn((_, cb) => cb({
                        origin: 'https://google.com',
                        data: {
                            template: 'Test template',
                        },
                    }));

                    app = mount(
                        <App composeView={ composeView }/>,
                    );
                });

                it('should not insert the template into the view', () => {
                    expect(composeView.insertTextIntoBodyAtCursor).not.toHaveBeenCalled();
                });
            });
        });
    });

    describe('#componentWillUnmount', () => {
        beforeEach(() => {
            global.window.addEventListener = jest.fn((_, cb) => cb({
                origin: 'https://email-snippets.herokuapp.com',
                data: {
                    template: 'Test template',
                },
            }));

            app = mount(
                <App composeView={ composeView }/>,
            );

            app.unmount();
        });

        it('should remove the event listener "message"', () => {
            expect(global.window.removeEventListener).toHaveBeenCalledWith(
                'message',
                expect.any(Function),
            );
        });
    });
});
