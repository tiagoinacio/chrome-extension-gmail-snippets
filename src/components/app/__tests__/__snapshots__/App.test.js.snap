// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App when iframe is loaded should not render the loader 1`] = `
<App
  composeView={
    Object {
      "insertTextIntoBodyAtCursor": [MockFunction],
    }
  }
>
  <div
    className="container"
  >
    <iframe
      className="iframe"
      onLoad={[Function]}
      src="https://email-snippets.herokuapp.com"
    >
      Your browser does not support iframes.
    </iframe>
  </div>
</App>
`;

exports[`App while iframe is not loaded should render the loader 1`] = `
<App
  composeView={
    Object {
      "insertTextIntoBodyAtCursor": [MockFunction],
    }
  }
>
  <div
    className="container"
  >
    <div
      className="loader"
    />
    <iframe
      className="iframe"
      onLoad={[Function]}
      src="https://email-snippets.herokuapp.com"
    >
      Your browser does not support iframes.
    </iframe>
  </div>
</App>
`;
